crypto/s2n_hkdf.c

```diff
@@ -126,7 +126,11 @@ static int s2n_libcrypto_hkdf_extract(struct s2n_hmac_state *hmac, s2n_hmac_algo
     const EVP_MD *digest = NULL;
     POSIX_GUARD_RESULT(s2n_hmac_md_from_alg(alg, &digest));
 
-    size_t pseudo_rand_key_len = (size_t) pseudo_rand_key->size;
+    size_t hmac_size = EVP_MD_size(digest);
+    POSIX_ENSURE_GT(hmac_size, 0);
+    POSIX_ENSURE(hmac_size <= pseudo_rand_key->size, S2N_ERR_HKDF_OUTPUT_SIZE);
+
+    size_t pseudo_rand_key_len = 0;
```

Comment:
Instead of duping can we add the validation to [s2n_hkdf_extract](https://github.com/aws/s2n-tls/blob/4f9124f21e5244516df9bf783d4675e6fc17a7c5/crypto/s2n_hkdf.c#L212). 

For example now this check has `POSIX_ENSURE_GT(hmac_size, 0);` but the other doesnt. Also POSIX_ENSURE_GT(hmac_size, 0); should also return `S2N_ERR_HKDF_OUTPUT_SIZE`.

---

Similarly are there common checks we can move to `s2n_hkdf` and `s2n_hkdf_expand` rather than the two impls?

Reply:
We could pull it out, but I left it separate so our custom implementation can call our own function to get the digest size (`s2n_hmac_digest_size`). This should be the same as the libcrypto version, but it seemed more correct to have the libcrypto version use its own digest size function. But if that doesn't make sense and it seems more clear to move it out, let me know.

The hmac_size > 0 check is because `EVP_MD_size` can't indicate a failure like`s2n_hmac_digest_size` can, so I wanted to make sure it was set to something reasonable. I don't expect this to ever fail, though, so I'm not sure this would indicate an error in the output size. S2N_ERR_SAFETY seems reasonable for this, but let me know if you disagree.

All of the checks that are the same are already moved out to the parent call: https://github.com/aws/s2n-tls/blob/c9dd66ed11b44e3b20d68883b60affc83a391bfc/crypto/s2n_hkdf.c#L211-L214

Reply:
I moved the check out to the parent function.

I looked at our `s2n_hmac_digest_size` implementation, and we pull the constants straight from the libcrypto header, so I'm less concerned about using this function for this now.